File: python/src/nr_ai_agent/metrics/multimodal.py
"""Multi-modal input detection — images, PDFs, audio, video in messages."""

import base64
import logging
import math
import struct
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)


def _get_image_dimensions_from_base64(b64_data: str) -> Optional[tuple]:
    """Return (width, height) from base64 image data, or None if unknown."""
    try:
        raw = base64.b64decode(b64_data, validate=False)

        # PNG: magic bytes + IHDR chunk at offset 8 — width/height at bytes 16-24
        if b64_data.startswith("iVBORw0KGgo") and len(raw) >= 24:
            (width,) = struct.unpack_from(">I", raw, 16)
            (height,) = struct.unpack_from(">I", raw, 20)
            return (width, height)

        # JPEG / WebP: rough estimate from file size
        if b64_data.startswith("/9j/") or b64_data.startswith("UklGR"):
            estimated_pixels = math.sqrt((len(raw) * 8) / 24)
            side = round(estimated_pixels)
            return (side, side)

        return None
    except Exception:
        return None


def _estimate_image_tokens(width: int, height: int) -> int:
    return math.ceil((width * height) / 750)


def _get_image_token_estimate(image_data: Dict[str, Any]) -> int:
    width = image_data.get("width")
    height = image_data.get("height")
    if isinstance(width, (int, float)) and isinstance(height, (int, float)):
        return _estimate_image_tokens(int(width), int(height))

    b64_data: Optional[str] = image_data.get("base64")
    source = image_data.get("source")
    if not b64_data and isinstance(source, dict) and source.get("type") == "base64":
        b64_data = source.get("data")

    if b64_data and isinstance(b64_data, str):
        dims = _get_image_dimensions_from_base64(b64_data)
        if dims:
            return _estimate_image_tokens(dims[0], dims[1])
        binary_size = (len(b64_data) * 3) / 4
        estimated_pixels = binary_size / 4
        estimated_dim = math.sqrt(estimated_pixels)
        return _estimate_image_tokens(round(estimated_dim), round(estimated_dim))

    logger.warning("Unable to determine image dimensions, using default estimate")
    return 512


def _parse_pdf_page_count(metadata: Optional[Dict[str, Any]]) -> int:
    if metadata and isinstance(metadata.get("pages"), (int, float)):
        return int(metadata["pages"])
    return 1


def _parse_audio_seconds(metadata: Optional[Dict[str, Any]]) -> float:
    if metadata:
        if isinstance(metadata.get("duration_seconds"), (int, float)):
            return float(metadata["duration_seconds"])
        if isinstance(metadata.get("duration"), (int, float)):
            return float(metadata["duration"])
    return 0.0


def detect_modalities(messages: List[Any]) -> Dict[str, Any]:
    """Detect input modalities from a message list (Anthropic or Gemini format)."""
    modalities: set = set()
    image_count = 0
    image_token_estimate = 0
    pdf_count = 0
    pdf_page_count = 0
    audio_seconds = 0.0
    video_seconds = 0.0
    text_tokens = 0

    if not isinstance(messages, list):
        modalities.add("text")
        return _build_result(modalities, image_count, image_token_estimate, pdf_count,
                              pdf_page_count, audio_seconds, video_seconds, text_tokens)

    for message in messages:
        if not message or not isinstance(message, dict):
            continue

        # Anthropic-style: content is a list of blocks
        if isinstance(message.get("content"), list):
            for block in message["content"]:
                if not isinstance(block, dict):
                    continue
                if block.get("type") == "text" and isinstance(block.get("text"), str):
                    modalities.add("text")
                    text_tokens += math.ceil(len(block["text"]) / 4)
                elif block.get("type") == "image":
                    modalities.add("image")
                    image_count += 1
                    image_token_estimate += _get_image_token_estimate(block)
                elif block.get("type") == "document":
                    modalities.add("pdf")
                    pdf_count += 1
                    pdf_page_count += _parse_pdf_page_count(block.get("metadata"))

        # Gemini-style: parts array
        if isinstance(message.get("parts"), list):
            for part in message["parts"]:
                if not isinstance(part, dict):
                    continue
                if isinstance(part.get("text"), str):
                    modalities.add("text")
                    text_tokens += math.ceil(len(part["text"]) / 4)

                inline_data = part.get("inlineData")
                if isinstance(inline_data, dict):
                    mime_type = inline_data.get("mimeType", "")
                    if mime_type.startswith("image/"):
                        modalities.add("image")
                        image_count += 1
                        data = inline_data.get("data")
                        if isinstance(data, str):
                            image_token_estimate += _get_image_token_estimate({"base64": data})
                        else:
                            image_token_estimate += 512
                    elif mime_type.startswith("audio/"):
                        modalities.add("audio")
                        audio_seconds += _parse_audio_seconds(inline_data)
                    elif mime_type.startswith("video/"):
                        modalities.add("video")
                        video_seconds += _parse_audio_seconds(inline_data)

                file_data = part.get("fileData")
                if isinstance(file_data, dict):
                    mime_type = file_data.get("mimeType", "")
                    if mime_type.startswith("image/"):
                        modalities.add("image")
                        image_count += 1
                        image_token_estimate += 512
                    elif mime_type.startswith("application/pdf"):
                        modalities.add("pdf")
                        pdf_count += 1
                        pdf_page_count += _parse_pdf_page_count(file_data)
                    elif mime_type.startswith("audio/"):
                        modalities.add("audio")
                        audio_seconds += _parse_audio_seconds(file_data)
                    elif mime_type.startswith("video/"):
                        modalities.add("video")
                        video_seconds += _parse_audio_seconds(file_data)

    if not modalities:
        modalities.add("text")

    return _build_result(modalities, image_count, image_token_estimate, pdf_count,
                          pdf_page_count, audio_seconds, video_seconds, text_tokens)


def _build_result(
    modalities: set,
    image_count: int,
    image_token_estimate: int,
    pdf_count: int,
    pdf_page_count: int,
    audio_seconds: float,
    video_seconds: float,
    text_tokens: int,
) -> Dict[str, Any]:
    return {
        "input_modalities": sorted(modalities),
        "image_count": image_count,
        "image_token_estimate": image_token_estimate,
        "pdf_count": pdf_count,
        "pdf_page_count": pdf_page_count,
        "audio_seconds": audio_seconds,
        "video_seconds": video_seconds,
        "text_tokens": text_tokens,
    }


def modality_metrics_to_custom_attributes(metrics: Dict[str, Any]) -> Dict[str, Any]:
    """Convert modality metrics to flat NR custom attribute dict."""
    return {
        "ai.input.modalities": ",".join(metrics["input_modalities"]),
        "ai.input.image_count": metrics["image_count"],
        "ai.input.image_token_estimate": metrics["image_token_estimate"],
        "ai.input.pdf_count": metrics["pdf_count"],
        "ai.input.pdf_page_count": metrics["pdf_page_count"],
        "ai.input.audio_seconds": round(metrics["audio_seconds"] * 100) / 100,
        "ai.input.video_seconds": round(metrics["video_seconds"] * 100) / 100,
        "ai.input.text_tokens": metrics["text_tokens"],
    }

File: packages/nr-ai-agent/src/metrics/multimodal.ts
import { createLogger } from '@nr-ai-observatory/shared';

const logger = createLogger('multimodal-tracker');

export interface MultiModalMetrics {
  readonly inputModalities: string[];
  readonly imageCount: number;
  readonly imageTokenEstimate: number;
  readonly pdfCount: number;
  readonly pdfPageCount: number;
  readonly audioSeconds: number;
  readonly videoSeconds: number;
  readonly textTokens: number;
}

// Helper: detect image dimensions from base64 data
function getImageDimensionsFromBase64(base64Data: string): { width: number; height: number } | null {
  try {
    // PNG: bytes 16-24 contain width/height as big-endian 32-bit ints
    if (base64Data.startsWith('iVBORw0KGgo')) {
      const binaryString = Buffer.from(base64Data, 'base64').toString('binary');
      if (binaryString.length >= 24) {
        const width = (binaryString.charCodeAt(16) << 24) | (binaryString.charCodeAt(17) << 16) |
                      (binaryString.charCodeAt(18) << 8) | binaryString.charCodeAt(19);
        const height = (binaryString.charCodeAt(20) << 24) | (binaryString.charCodeAt(21) << 16) |
                       (binaryString.charCodeAt(22) << 8) | binaryString.charCodeAt(23);
        return { width: width >>> 0, height: height >>> 0 };
      }
    }

    // JPEG: look for SOF marker (0xFFC0-0xFFC9) which contains dimensions
    if (base64Data.startsWith('/9j/')) {
      const buffer = Buffer.from(base64Data, 'base64');
      // Simplified: just estimate from file size
      // Real JPEG parsing would decode SOF marker
      const estimatedPixels = Math.sqrt((buffer.length * 8) / 24); // rough estimate
      return { width: Math.round(estimatedPixels), height: Math.round(estimatedPixels) };
    }

    // WebP: similar approach
    if (base64Data.startsWith('UklGR')) {
      const buffer = Buffer.from(base64Data, 'base64');
      const estimatedPixels = Math.sqrt((buffer.length * 8) / 32);
      return { width: Math.round(estimatedPixels), height: Math.round(estimatedPixels) };
    }

    return null;
  } catch {
    return null;
  }
}

// Helper: estimate image tokens using Anthropic's formula
function estimateImageTokens(width: number, height: number): number {
  return Math.ceil((width * height) / 750);
}

// Helper: detect image dimensions or estimate from base64 size
function getImageTokenEstimate(imageData: Record<string, unknown>): number {
  const width = imageData.width as number | undefined;
  const height = imageData.height as number | undefined;

  // If dimensions are provided, use them
  if (width && height) {
    return estimateImageTokens(width, height);
  }

  // Try to extract base64 data
  let base64Data = imageData.base64 as string | undefined;
  const source = imageData.source as Record<string, unknown> | undefined;

  if (!base64Data && source?.type === 'base64' && source?.data) {
    base64Data = source.data as string;
  }

  if (base64Data) {
    // Try to detect dimensions from base64 header
    const dims = getImageDimensionsFromBase64(base64Data);
    if (dims) {
      return estimateImageTokens(dims.width, dims.height);
    }

    // Fallback: estimate from base64 size
    // Base64 encoding is ~4/3 larger; assume ~4 bytes per pixel
    const binarySize = (base64Data.length * 3) / 4;
    const estimatedPixels = binarySize / 4;
    const estimatedDim = Math.sqrt(estimatedPixels);
    return estimateImageTokens(Math.round(estimatedDim), Math.round(estimatedDim));
  }

  // Default estimate for images without size info
  logger.warn('Unable to determine image dimensions, using default estimate', {
    hasBase64: !!imageData.base64,
    hasSource: !!imageData.source,
  });
  return 512; // Conservative default
}

// Helper: parse PDF page count from metadata or filename
function parsePdfPageCount(metadata?: Record<string, unknown>): number {
  if (metadata?.pages && typeof metadata.pages === 'number') {
    return metadata.pages;
  }
  return 1; // Default to 1 if unknown
}

// Helper: parse audio duration in seconds
function parseAudioSeconds(metadata?: Record<string, unknown>): number {
  if (metadata?.duration_seconds && typeof metadata.duration_seconds === 'number') {
    return metadata.duration_seconds;
  }
  if (metadata?.duration && typeof metadata.duration === 'number') {
    return metadata.duration;
  }
  return 0;
}

// Main detection function - generic message format
export function detectModalities(messages: unknown[]): MultiModalMetrics {
  const modalities = new Set<string>();
  let imageCount = 0;
  let imageTokenEstimate = 0;
  let pdfCount = 0;
  let pdfPageCount = 0;
  let audioSeconds = 0;
  let videoSeconds = 0;
  let textTokens = 0;

  if (!Array.isArray(messages)) {
    modalities.add('text');
    return {
      inputModalities: Array.from(modalities),
      imageCount,
      imageTokenEstimate,
      pdfCount,
      pdfPageCount,
      audioSeconds,
      videoSeconds,
      textTokens,
    };
  }

  for (const message of messages) {
    if (!message || typeof message !== 'object') continue;

    const msg = message as Record<string, unknown>;

    // Handle Anthropic-style messages (content array with blocks)
    if (Array.isArray(msg.content)) {
      for (const block of msg.content) {
        if (!block || typeof block !== 'object') continue;

        const contentBlock = block as Record<string, unknown>;

        if (contentBlock.type === 'text' && typeof contentBlock.text === 'string') {
          modalities.add('text');
          textTokens += Math.ceil(contentBlock.text.length / 4);
        }

        if (contentBlock.type === 'image') {
          modalities.add('image');
          imageCount += 1;
          imageTokenEstimate += getImageTokenEstimate(contentBlock);
        }

        if (contentBlock.type === 'document') {
          modalities.add('pdf');
          pdfCount += 1;
          const pdfMeta = contentBlock.metadata as Record<string, unknown> | undefined;
          pdfPageCount += parsePdfPageCount(pdfMeta);
        }
      }
    }

    // Handle Gemini-style messages (parts array)
    if (Array.isArray(msg.parts)) {
      for (const part of msg.parts) {
        if (!part || typeof part !== 'object') continue;

        const partData = part as Record<string, unknown>;

        if (partData.text && typeof partData.text === 'string') {
          modalities.add('text');
          textTokens += Math.ceil(partData.text.length / 4);
        }

        // Gemini inlineData
        if (partData.inlineData && typeof partData.inlineData === 'object') {
          const inlineData = partData.inlineData as Record<string, unknown>;
          const mimeType = typeof inlineData.mimeType === 'string' ? inlineData.mimeType : '';

          if (mimeType.startsWith('image/')) {
            modalities.add('image');
            imageCount += 1;
            // Estimate from data field
            if (typeof inlineData.data === 'string') {
              imageTokenEstimate += getImageTokenEstimate({ base64: inlineData.data });
            } else {
              imageTokenEstimate += 512; // Conservative default
            }
          }

          if (mimeType.startsWith('audio/')) {
            modalities.add('audio');
            audioSeconds += parseAudioSeconds(inlineData);
          }

          if (mimeType.startsWith('video/')) {
            modalities.add('video');
            videoSeconds += parseAudioSeconds(inlineData);
          }
        }

        // Gemini fileData
        if (partData.fileData && typeof partData.fileData === 'object') {
          const fileData = partData.fileData as Record<string, unknown>;
          const mimeType = typeof fileData.mimeType === 'string' ? fileData.mimeType : '';

          if (mimeType.startsWith('image/')) {
            modalities.add('image');
            imageCount += 1;
            imageTokenEstimate += 512; // Default for uploaded files
          }

          if (mimeType.startsWith('application/pdf')) {
            modalities.add('pdf');
            pdfCount += 1;
            pdfPageCount += parsePdfPageCount(fileData);
          }

          if (mimeType.startsWith('audio/')) {
            modalities.add('audio');
            audioSeconds += parseAudioSeconds(fileData);
          }

          if (mimeType.startsWith('video/')) {
            modalities.add('video');
            videoSeconds += parseAudioSeconds(fileData);
          }
        }
      }
    }
  }

  // Always include text if no other modalities detected
  if (modalities.size === 0) {
    modalities.add('text');
  }

  return {
    inputModalities: Array.from(modalities).sort(),
    imageCount,
    imageTokenEstimate,
    pdfCount,
    pdfPageCount,
    audioSeconds,
    videoSeconds,
    textTokens,
  };
}

export function modalityMetricsToCustomAttributes(metrics: MultiModalMetrics): Record<string, string | number> {
  const attrs: Record<string, string | number> = {
    'ai.input.modalities': metrics.inputModalities.join(','),
    'ai.input.image_count': metrics.imageCount,
    'ai.input.image_token_estimate': metrics.imageTokenEstimate,
    'ai.input.pdf_count': metrics.pdfCount,
    'ai.input.pdf_page_count': metrics.pdfPageCount,
    'ai.input.audio_seconds': Math.round(metrics.audioSeconds * 100) / 100,
    'ai.input.video_seconds': Math.round(metrics.videoSeconds * 100) / 100,
    'ai.input.text_tokens': metrics.textTokens,
  };

  return attrs;
}
